// @sudoku/stores/gameSession.js
import { SUDOKU_SIZE } from '@sudoku/constants';
import { derived, get, writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';
import { grid } from './grid';
import { hints } from './hints';

function emptyGrid() {
	return Array.from({ length: SUDOKU_SIZE }, () => Array(SUDOKU_SIZE).fill(0));
}

function warnHintNotPossible() {
	if (
		typeof window !== 'undefined' &&
		typeof window.dispatchEvent === 'function' &&
		typeof window.CustomEvent === 'function'
	) {
		window.dispatchEvent(new window.CustomEvent('sudoku:hint-not-possible'));
	}
}

function createGameSession() {
	const gameStore = writable(createGame({ sudoku: createSudoku(emptyGrid()) }));

	grid.subscribe(($grid) => {
		if (!$grid) return;
		gameStore.set(createGame({ sudoku: createSudoku($grid) }));
	});

	const publish = () => gameStore.set(get(gameStore));

	const apply = (mutator) => {
		const game = get(gameStore);
		const result = mutator(game);
		publish();
		return result;
	};

	const activeGrid = derived(gameStore, ($g) => ($g ? $g.getSudoku().getGrid() : emptyGrid()));
	const canUndo = derived(gameStore, ($g) => !!$g && $g.canUndo());
	const canRedo = derived(gameStore, ($g) => !!$g && $g.canRedo());
	const isExploring = derived(gameStore, ($g) => !!$g && $g.isExploring());
	const exploreFailure = derived(gameStore, ($g) => ($g ? $g.getExploreFailure() : null));
	const invalidCells = derived(gameStore, ($g) => {
		if (!$g) return [];
		return $g.getConflictingCells().map(({ row, col }) => `${col},${row}`);
	});
	const exploreBranches = derived(gameStore, ($g) => ($g ? $g.listBranches() : []));
	const currentBranchId = derived(gameStore, ($g) => ($g ? $g.getCurrentBranchId() : 'main'));

	return {
		subscribe: activeGrid.subscribe,
		canUndo,
		canRedo,
		isExploring,
		exploreFailure,
		invalidCells,
		exploreBranches,
		currentBranchId,

		set: (pos, value) =>
			apply((game) => game.guess({ row: pos.y, col: pos.x, value })),

		undo: () => apply((game) => game.undo()),
		redo: () => apply((game) => game.redo()),

		applyHint: (pos) =>
			apply((game) => {
				const hint = game.getNextHint();
				if (!hint) {
					warnHintNotPossible();
					return false;
				}
				const accepted = game.guess(hint);
				if (accepted) {
					hints.useHint();
				}
				return accepted;
			}),

		peekHint: () => {
			const game = get(gameStore);
			return game ? game.peekNextHint() : null;
		},

		startExplore: () => apply((game) => game.startExplore()),
		commitExplore: () => apply((game) => game.commitExplore()),
		discardExplore: () => apply((game) => game.discardExplore()),
		backtrackExplore: () => apply((game) => game.backtrackExplore()),
		forkBranch: () => apply((game) => game.forkBranch()),
		switchBranch: (id) => apply((game) => game.switchBranch(id)),
	};
}

export const gameSession = createGameSession();
export const canUndo = gameSession.canUndo;
export const canRedo = gameSession.canRedo;
export const isExploring = gameSession.isExploring;
export const exploreFailure = gameSession.exploreFailure;
export const invalidCells = gameSession.invalidCells;
export const exploreBranches = gameSession.exploreBranches;
export const currentBranchId = gameSession.currentBranchId;
